src/useRouteQuery/useRouteQuery.ts

```diff
@@ -0,0 +1,112 @@
+import { reactive, watch } from 'vue'
+import { LocationQuery, LocationQueryValue, useRoute, useRouter } from 'vue-router'
+
+type UseRouteQuery = {
+  query: LocationQuery,
+  clear: () => void,
+  set: (key: string, value: string) => void,
+  get: (key: string) => LocationQueryValue | LocationQueryValue[],
+  remove: (key: string) => void,
+}
+
+type QueryOperation = (query: LocationQuery) => LocationQuery
+
+export function useRouteQuery(): UseRouteQuery {
+  const route = useRoute()
+  const router = useRouter()
+  const operations = reactive<QueryOperation[]>([])
+  let interval: ReturnType<typeof setTimeout>
+
+  const clear = (): void => {
+    router.push({ query: {} })
+  }
+
+  const set = (key: string, value: string): void => {
+    query[key] = value
+  }
+
+  const get = (key: string): LocationQueryValue | LocationQueryValue[] => {
+    return query[key]
+  }
+
+  const remove = (key: string): void => {
+    delete query[key]
+  }
+
+  const query: LocationQuery = new Proxy({}, {
+    get(target, property) {
+      if (typeof property === 'symbol') {
+        return Reflect.get(target, property)
+      }
+
+      const query = applyQueryOperations(route.query, operations)
+
+      return query[property]
+    },
+    set(target, property, value) {
+      if (typeof property === 'symbol' || typeof value !== 'string') {
+        return Reflect.set(target, property, value)
+      }
+
+      operations.push(updateOperation(property, value))
+
+      return true
+    },
+    deleteProperty(target, property) {
+      if (typeof property === 'symbol') {
+        return Reflect.deleteProperty(target, property)
+      }
+
+      operations.push(deleteOperation(property))
+
+      return true
+    },
+  })
+
+  const update = (): void => {
+    const query = applyQueryOperations(route.query, operations)
+
+    router.push({ query })
+  }
+
+  watch(operations, () => {
+    clearInterval(interval)
+    interval = setTimeout(update)
+  })
+
+  return {
+    clear,
+    set,
+    get,
+    remove,
+    query,
+  }
+}
+
+function applyQueryOperations(currentQuery: LocationQuery, operations: QueryOperation[]): LocationQuery {
+  let query: LocationQuery = { ...currentQuery }
+
+  for (const operation of operations) {
+    query = operation(query)
+  }
+
+  return query
+}
+
+function updateOperation(key: string, value: string): QueryOperation {
```

Comment:
nit, this function name isn't super clear to me. As is, I think it's more like `getUpdateOperation`, but if it were me, I might also take the `operations` as a param and call it `queueUpdateOperation`, feels like it works nicely to `queue` operations, and then eventually `apply` operations?

Reply:
I don't want to over think this considering these are private methods that are not exported. I updated the naming of the operation methods to be `updateOperationFactory` and `deleteOperationFactory` which accurately describes what they do. I also broke out the creation of an operation from the `operations.push` which makes it more readable IMO.